In [1]:
import pandas as pd
import numpy as np
from faker import Faker

### Training, Section and Question

In [10]:
question_address = []
for t in [1,2]:
    for s in [1,2,3,4]:
        for sub in [1,2,3,4,5]:
            for q in [1,2,3,4,5,6]:
                for a in [1,2,3]:
                    question_address.append("Tra{}_Sec{}_Sub{}_Qus{}_Att{}".format(t,s,sub,q,a))

question_address.insert(0,"user_name") 

In [11]:
df_first_time = pd.DataFrame(columns=question_address)

In [12]:
df_first_time.head(1)

Empty DataFrame
Columns: [user_name, Tra1_Sec1_Sub1_Qus1_Att1, Tra1_Sec1_Sub1_Qus1_Att2, Tra1_Sec1_Sub1_Qus1_Att3, Tra1_Sec1_Sub1_Qus2_Att1, Tra1_Sec1_Sub1_Qus2_Att2, Tra1_Sec1_Sub1_Qus2_Att3, Tra1_Sec1_Sub1_Qus3_Att1, Tra1_Sec1_Sub1_Qus3_Att2, Tra1_Sec1_Sub1_Qus3_Att3, Tra1_Sec1_Sub1_Qus4_Att1, Tra1_Sec1_Sub1_Qus4_Att2, Tra1_Sec1_Sub1_Qus4_Att3, Tra1_Sec1_Sub1_Qus5_Att1, Tra1_Sec1_Sub1_Qus5_Att2, Tra1_Sec1_Sub1_Qus5_Att3, Tra1_Sec1_Sub1_Qus6_Att1, Tra1_Sec1_Sub1_Qus6_Att2, Tra1_Sec1_Sub1_Qus6_Att3, Tra1_Sec1_Sub2_Qus1_Att1, Tra1_Sec1_Sub2_Qus1_Att2, Tra1_Sec1_Sub2_Qus1_Att3, Tra1_Sec1_Sub2_Qus2_Att1, Tra1_Sec1_Sub2_Qus2_Att2, Tra1_Sec1_Sub2_Qus2_Att3, Tra1_Sec1_Sub2_Qus3_Att1, Tra1_Sec1_Sub2_Qus3_Att2, Tra1_Sec1_Sub2_Qus3_Att3, Tra1_Sec1_Sub2_Qus4_Att1, Tra1_Sec1_Sub2_Qus4_Att2, Tra1_Sec1_Sub2_Qus4_Att3, Tra1_Sec1_Sub2_Qus5_Att1, Tra1_Sec1_Sub2_Qus5_Att2, Tra1_Sec1_Sub2_Qus5_Att3, Tra1_Sec1_Sub2_Qus6_Att1, Tra1_Sec1_Sub2_Qus6_Att2, Tra1_Sec1_Sub2_Qus6_Att3, Tra1_Sec1_Sub3_Qus1_Att1, Tra1_Sec1_Sub3_Qus1_Att2, Tra1_Sec1_Sub3_Qus1_Att3, Tra1_Sec1_Sub3_Qus2_Att1, Tra1_Sec1_Sub3_Qus2_Att2, Tra1_Sec1_Sub3_Qus2_Att3, Tra1_Sec1_Sub3_Qus3_Att1, Tra1_Sec1_Sub3_Qus3_Att2, Tra1_Sec1_Sub3_Qus3_Att3, Tra1_Sec1_Sub3_Qus4_Att1, Tra1_Sec1_Sub3_Qus4_Att2, Tra1_Sec1_Sub3_Qus4_Att3, Tra1_Sec1_Sub3_Qus5_Att1, Tra1_Sec1_Sub3_Qus5_Att2, Tra1_Sec1_Sub3_Qus5_Att3, Tra1_Sec1_Sub3_Qus6_Att1, Tra1_Sec1_Sub3_Qus6_Att2, Tra1_Sec1_Sub3_Qus6_Att3, Tra1_Sec1_Sub4_Qus1_Att1, Tra1_Sec1_Sub4_Qus1_Att2, Tra1_Sec1_Sub4_Qus1_Att3, Tra1_Sec1_Sub4_Qus2_Att1, Tra1_Sec1_Sub4_Qus2_Att2, Tra1_Sec1_Sub4_Qus2_Att3, Tra1_Sec1_Sub4_Qus3_Att1, Tra1_Sec1_Sub4_Qus3_Att2, Tra1_Sec1_Sub4_Qus3_Att3, Tra1_Sec1_Sub4_Qus4_Att1, Tra1_Sec1_Sub4_Qus4_Att2, Tra1_Sec1_Sub4_Qus4_Att3, Tra1_Sec1_Sub4_Qus5_Att1, Tra1_Sec1_Sub4_Qus5_Att2, Tra1_Sec1_Sub4_Qus5_Att3, Tra1_Sec1_Sub4_Qus6_Att1, Tra1_Sec1_Sub4_Qus6_Att2, Tra1_Sec1_Sub4_Qus6_Att3, Tra1_Sec1_Sub5_Qus1_Att1, Tra1_Sec1_Sub5_Qus1_Att2, Tra1_Sec1_Sub5_Qus1_Att3, Tra1_Sec1_Sub5_Qus2_Att1, Tra1_Sec1_Sub5_Qus2_Att2, Tra1_Sec1_Sub5_Qus2_Att3, Tra1_Sec1_Sub5_Qus3_Att1, Tra1_Sec1_Sub5_Qus3_Att2, Tra1_Sec1_Sub5_Qus3_Att3, Tra1_Sec1_Sub5_Qus4_Att1, Tra1_Sec1_Sub5_Qus4_Att2, Tra1_Sec1_Sub5_Qus4_Att3, Tra1_Sec1_Sub5_Qus5_Att1, Tra1_Sec1_Sub5_Qus5_Att2, Tra1_Sec1_Sub5_Qus5_Att3, Tra1_Sec1_Sub5_Qus6_Att1, Tra1_Sec1_Sub5_Qus6_Att2, Tra1_Sec1_Sub5_Qus6_Att3, Tra1_Sec2_Sub1_Qus1_Att1, Tra1_Sec2_Sub1_Qus1_Att2, Tra1_Sec2_Sub1_Qus1_Att3, Tra1_Sec2_Sub1_Qus2_Att1, Tra1_Sec2_Sub1_Qus2_Att2, Tra1_Sec2_Sub1_Qus2_Att3, Tra1_Sec2_Sub1_Qus3_Att1, Tra1_Sec2_Sub1_Qus3_Att2, Tra1_Sec2_Sub1_Qus3_Att3, ...]
Index: []

[0 rows x 721 columns]

### User names

In [13]:
fake = Faker('en_IN')


In [14]:
names = []
for _ in range(1008):
    names.append(fake.name())


names = list(set(names))

In [30]:
df_first_time['user_name'] = names

In [51]:
df_first_time.head(1)

user_name Tra1_Sec1_Sub1_Qus1_Att1 Tra1_Sec1_Sub1_Qus1_Att2  \
0  Emir Kumar                        0                        0   

  Tra1_Sec1_Sub1_Qus1_Att3 Tra1_Sec1_Sub1_Qus2_Att1 Tra1_Sec1_Sub1_Qus2_Att2  \
0                        0                        1                        1   

  Tra1_Sec1_Sub1_Qus2_Att3 Tra1_Sec1_Sub1_Qus3_Att1 Tra1_Sec1_Sub1_Qus3_Att2  \
0                        0                        0                        0   

  Tra1_Sec1_Sub1_Qus3_Att3  ... Tra2_Sec4_Sub5_Qus3_Att3  \
0                        0  ...                        0   

  Tra2_Sec4_Sub5_Qus4_Att1 Tra2_Sec4_Sub5_Qus4_Att2 Tra2_Sec4_Sub5_Qus4_Att3  \
0                        0                        1                        0   

  Tra2_Sec4_Sub5_Qus5_Att1 Tra2_Sec4_Sub5_Qus5_Att2 Tra2_Sec4_Sub5_Qus5_Att3  \
0                        0                        0                        0   

  Tra2_Sec4_Sub5_Qus6_Att1 Tra2_Sec4_Sub5_Qus6_Att2 Tra2_Sec4_Sub5_Qus6_Att3  
0                        0                        0                        0  

[1 rows x 721 columns]

### First time score

In [53]:
%%time
# ToDo (Shivam): Make this fast

for i in np.arange(0,df_first_time.shape[0]):
    print(i)  # ToDo(Shivam): Replace this print statement with a percentage compeltion number updating on it's own
    for col in df_first_time.columns[1:]:
        if i <= round(df_first_time.shape[0]/4):
            if np.random.normal(0.4,0.2)>0.5:
                df_first_time.loc[i,col] = 1
            else:
                df_first_time.loc[i,col] = 0
        elif i <= round(df_first_time.shape[0]/2):
            if np.random.normal(0.8,0.2)>0.5:
                df_first_time.loc[i,col] = 1
            else:
                df_first_time.loc[i,col] = 0
        elif i <= (df_first_time.shape[0]-round(df_first_time.shape[0]/4)):
            if np.random.normal(df_first_time.columns[1:].get_loc(col)/df_first_time.shape[1],0.2)>0.5:
                df_first_time.loc[i,col] = 1
            else:
                df_first_time.loc[i,col] = 0
        else:
            if np.random.normal(1-df_first_time.columns[1:].get_loc(col)/df_first_time.shape[1],0.2)>0.5:
                df_first_time.loc[i,col] = 1
            else:
                df_first_time.loc[i,col] = 0
            


0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
27

In [55]:
df_first_time.tail()

user_name Tra1_Sec1_Sub1_Qus1_Att1 Tra1_Sec1_Sub1_Qus1_Att2  \
1000        Rhea Lad                        1                        1   
1001    Yakshit Brar                        1                        1   
1002   Rohan Chhabra                        1                        1   
1003   Vaibhav Borra                        1                        1   
1004  Nayantara Yogi                        1                        1   

     Tra1_Sec1_Sub1_Qus1_Att3 Tra1_Sec1_Sub1_Qus2_Att1  \
1000                        1                        1   
1001                        1                        1   
1002                        1                        1   
1003                        1                        1   
1004                        1                        1   

     Tra1_Sec1_Sub1_Qus2_Att2 Tra1_Sec1_Sub1_Qus2_Att3  \
1000                        1                        1   
1001                        1                        1   
1002                        1                        1   
1003                        1                        1   
1004                        1                        1   

     Tra1_Sec1_Sub1_Qus3_Att1 Tra1_Sec1_Sub1_Qus3_Att2  \
1000                        1                        1   
1001                        1                        1   
1002                        1                        1   
1003                        1                        1   
1004                        1                        1   

     Tra1_Sec1_Sub1_Qus3_Att3  ... Tra2_Sec4_Sub5_Qus3_Att3  \
1000                        1  ...                        0   
1001                        1  ...                        0   
1002                        1  ...                        0   
1003                        1  ...                        0   
1004                        1  ...                        0   

     Tra2_Sec4_Sub5_Qus4_Att1 Tra2_Sec4_Sub5_Qus4_Att2  \
1000                        0                        0   
1001                        0                        0   
1002                        0                        0   
1003                        0                        0   
1004                        0                        0   

     Tra2_Sec4_Sub5_Qus4_Att3 Tra2_Sec4_Sub5_Qus5_Att1  \
1000                        0                        0   
1001                        0                        0   
1002                        0                        0   
1003                        0                        0   
1004                        0                        0   

     Tra2_Sec4_Sub5_Qus5_Att2 Tra2_Sec4_Sub5_Qus5_Att3  \
1000                        0                        0   
1001                        0                        0   
1002                        0                        0   
1003                        0                        0   
1004                        0                        0   

     Tra2_Sec4_Sub5_Qus6_Att1 Tra2_Sec4_Sub5_Qus6_Att2  \
1000                        0                        0   
1001                        0                        0   
1002                        0                        0   
1003                        0                        0   
1004                        0                        0   

     Tra2_Sec4_Sub5_Qus6_Att3  
1000                        0  
1001                        0  
1002                        0  
1003                        0  
1004                        0  

[5 rows x 721 columns]

### Writing data to Excel

In [1]:
# ToDo (Shivam): Make the version control dynamic by using latest datetime = DONE
writer = pd.ExcelWriter('df_first_time_v3_{}.xlsx'.format(pd.datetime.today().strftime('%d%m%y- %H %M %S'))) 
df_first_time.to_excel(writer,'Sheet1')
writer.save()

### Creating dataset 1) Training wise 2)Section wise 3)Sub-section wise

In [2]:
# ToDo(Shivam): Reading the above excel file to continue working (cant be done)
df_first_time = pd.read_excel('first_time_data_v3.xlsx')

In [3]:
df_first_time.head()

user_name  Tra1_Sec1_Sub1_Qus1_Att1  Tra1_Sec1_Sub1_Qus1_Att2  \
0         Emir Kumar                         0                         0   
1        Krish Bains                         0                         0   
2      Akarsh Suresh                         0                         0   
3    Priyansh Bhalla                         0                         0   
4  Darshit Sabharwal                         0                         0   

   Tra1_Sec1_Sub1_Qus1_Att3  Tra1_Sec1_Sub1_Qus2_Att1  \
0                         0                         0   
1                         0                         1   
2                         1                         0   
3                         0                         1   
4                         1                         0   

   Tra1_Sec1_Sub1_Qus2_Att2  Tra1_Sec1_Sub1_Qus2_Att3  \
0                         0                         1   
1                         1                         1   
2                         1                         1   
3                         0                         0   
4                         0                         0   

   Tra1_Sec1_Sub1_Qus3_Att1  Tra1_Sec1_Sub1_Qus3_Att2  \
0                         0                         0   
1                         1                         0   
2                         0                         0   
3                         0                         0   
4                         1                         0   

   Tra1_Sec1_Sub1_Qus3_Att3  ...  Tra2_Sec4_Sub5_Qus3_Att3  \
0                         0  ...                         0   
1                         1  ...                         1   
2                         0  ...                         0   
3                         0  ...                         1   
4                         0  ...                         0   

   Tra2_Sec4_Sub5_Qus4_Att1  Tra2_Sec4_Sub5_Qus4_Att2  \
0                         0                         1   
1                         1                         0   
2                         0                         0   
3                         1                         0   
4                         1                         1   

   Tra2_Sec4_Sub5_Qus4_Att3  Tra2_Sec4_Sub5_Qus5_Att1  \
0                         0                         0   
1                         0                         1   
2                         0                         1   
3                         0                         0   
4                         0                         0   

   Tra2_Sec4_Sub5_Qus5_Att2  Tra2_Sec4_Sub5_Qus5_Att3  \
0                         0                         0   
1                         0                         1   
2                         1                         0   
3                         0                         0   
4                         0                         0   

   Tra2_Sec4_Sub5_Qus6_Att1  Tra2_Sec4_Sub5_Qus6_Att2  \
0                         0                         1   
1                         0                         1   
2                         1                         0   
3                         0                         1   
4                         0                         0   

   Tra2_Sec4_Sub5_Qus6_Att3  
0                         1  
1                         0  
2                         0  
3                         0  
4                         0  

[5 rows x 721 columns]

### 1) Training wise

In [4]:
# Getting names or numbers of trainings
training_numbers = []
for col in df_first_time.columns:
    if 'Tra' in col:
        training_numbers.append(col[0:4])
        
training_names = sorted(list(set(training_numbers)))
training_names

['Tra1', 'Tra2']

In [5]:
# Getting score for users in each training and overall score
# Granularity - user_name

training_wise_score = pd.DataFrame(columns=['user_name'])
training_wise_score.user_name = df_first_time.user_name

for training in training_names:
    use_col=[]
    for col in df_first_time.columns:
        if training in col:
            use_col.append(col)
    training_wise_score[training+'_score'] = df_first_time.loc[:, use_col].sum(axis=1)
training_wise_score['Total_score'] = training_wise_score.iloc[:,1:].sum(axis=1)   
training_wise_score.set_index('user_name',inplace= True)
training_wise_score.head()
    


Tra1_score  Tra2_score  Total_score
user_name                                             
Emir Kumar                106         101          207
Krish Bains               105         107          212
Akarsh Suresh             112         126          238
Priyansh Bhalla           119          99          218
Darshit Sabharwal         111         107          218

### 2) Section wise

In [6]:
# Getting names of Sections in each training
section_wise=[]
for col in df_first_time.columns:
    if 'Tra1_Sec' in col or 'Tra2_Sec' in col :
        section_wise.append(col[0:9])
section_wise = sorted(list(set(section_wise)))
section_wise

['Tra1_Sec1',
 'Tra1_Sec2',
 'Tra1_Sec3',
 'Tra1_Sec4',
 'Tra2_Sec1',
 'Tra2_Sec2',
 'Tra2_Sec3',
 'Tra2_Sec4']

In [27]:
# Getting score for users in each training: Sections wise and overall score
# Granularity : users name

section_wise_score = pd.DataFrame(columns=['user_name'])
section_wise_score.user_name = df_first_time.user_name

for section in section_wise:
    sec_col =[]
    for col in df_first_time.columns:
        if section in col:
            sec_col.append(col)
    section_wise_score[section+ '_score'] = df_first_time.loc[:,sec_col].sum(axis=1)

total = section_wise_score.iloc[0:,1:].sum(axis=1)
section_wise_score['total'] = total
section_wise_score.set_index('user_name',inplace= True)
section_wise_score.tail()


Tra1_Sec1_score  Tra1_Sec2_score  Tra1_Sec3_score  \
user_name                                                           
Rhea Lad                     90               83               68   
Yakshit Brar                 90               86               74   
Rohan Chhabra                89               84               75   
Vaibhav Borra                90               83               69   
Nayantara Yogi               90               88               72   

                Tra1_Sec4_score  Tra2_Sec1_score  Tra2_Sec2_score  \
user_name                                                           
Rhea Lad                     52               32               14   
Yakshit Brar                 51               40               14   
Rohan Chhabra                57               37               13   
Vaibhav Borra                53               43               17   
Nayantara Yogi               55               25               17   

                Tra2_Sec3_score  Tra2_Sec4_score  total  
user_name                                                
Rhea Lad                      9                0    348  
Yakshit Brar                  3                0    358  
Rohan Chhabra                 7                0    362  
Vaibhav Borra                 7                1    363  
Nayantara Yogi                3                2    352

### 3) Sub-Section wise

In [8]:
# Getting the names of all Sub-sections: Training wise
sub_section = []
for col in df_first_time.columns:
    if 'Tra1_Sec1_Sub' in col or 'Tra2_Sec2_Sub' in col:
        sub_section.append(col[0:14])
sub_section = sorted(list(set(sub_section)))
sub_section

['Tra1_Sec1_Sub1',
 'Tra1_Sec1_Sub2',
 'Tra1_Sec1_Sub3',
 'Tra1_Sec1_Sub4',
 'Tra1_Sec1_Sub5',
 'Tra2_Sec2_Sub1',
 'Tra2_Sec2_Sub2',
 'Tra2_Sec2_Sub3',
 'Tra2_Sec2_Sub4',
 'Tra2_Sec2_Sub5']

In [9]:
# Getting score for users in each Training: Sub-Sections wise and overall score
# Granularity : users name
sub_section_wise = pd.DataFrame(columns=['user_name'])
sub_section_wise.user_name = df_first_time.user_name

for subsec in sub_section:
    sub_sec =[]
    for col in df_first_time.columns:
        if subsec in col:
            sub_sec.append(col)
            sub_section_wise[subsec] = df_first_time.loc[:,sub_sec].sum(axis=1)
total = sub_section_wise.iloc[0:,1:].sum(axis=1)
sub_section_wise['total'] = total
sub_section_wise.set_index('user_name',inplace= True)
sub_section_wise.head()


Tra1_Sec1_Sub1  Tra1_Sec1_Sub2  Tra1_Sec1_Sub3  \
user_name                                                           
Emir Kumar                      4               8               6   
Krish Bains                     6               4               5   
Akarsh Suresh                   7               5               7   
Priyansh Bhalla                 5               7               9   
Darshit Sabharwal               5               2               3   

                   Tra1_Sec1_Sub4  Tra1_Sec1_Sub5  Tra2_Sec2_Sub1  \
user_name                                                           
Emir Kumar                      1               3               3   
Krish Bains                     4               9               5   
Akarsh Suresh                   2               6               7   
Priyansh Bhalla                 8               5               5   
Darshit Sabharwal               4               4               5   

                   Tra2_Sec2_Sub2  Tra2_Sec2_Sub3  Tra2_Sec2_Sub4  \
user_name                                                           
Emir Kumar                      5               4               5   
Krish Bains                     6               7               6   
Akarsh Suresh                   3               7               6   
Priyansh Bhalla                 3               6               3   
Darshit Sabharwal               3               4               7   

                   Tra2_Sec2_Sub5  total  
user_name                                 
Emir Kumar                      4     43  
Krish Bains                     4     56  
Akarsh Suresh                   6     56  
Priyansh Bhalla                 5     56  
Darshit Sabharwal               5     42

## Exporting the Scores data file to Excel having datetime stamp

In [77]:
# Exporting Scores: Training,section,sub-section wise

score = {'Training':training_wise_score,'Section':section_wise_score,'Sub-section':sub_section_wise}
for key, i in score.items():
    writer = pd.ExcelWriter('users_score_'+key+'_{}.xlsx'.format(pd.datetime.today().strftime('%d%m%y- %H %M %S')))
    i.to_excel(writer,'Sheet1')
    writer.save()
    print("All files executated Succesfully")

<ipython-input-77-a3c4c8ada4f5>:3: FutureWarning: The pandas.datetime class is deprecated and will be removed from pandas in a future version. Import from datetime module instead.
  writer = pd.ExcelWriter('users_score_'+key+'_{}.xlsx'.format(pd.datetime.today().strftime('%d%m%y- %H %M %S')))
All files executated Succesfully
All files executated Succesfully
All files executated Succesfully


# Visualizing the data

In [10]:
import plotly.express as px
import plotly.graph_objs as go
import plotly.offline as pyo
import matplotlib.pyplot as plt 


#### 1. Line chart for all 4 types of players (best, worst, improving, deteriorating) over time

In [66]:
section_wise_score_trans = section_wise_score.T
section_wise_score_trans.head()

user_name        Emir Kumar  Krish Bains  Akarsh Suresh  Priyansh Bhalla  \
Tra1_Sec1_score          22           28             27               34   
Tra1_Sec2_score          25           28             30               21   
Tra1_Sec3_score          31           28             27               38   
Tra1_Sec4_score          28           21             28               26   
Tra2_Sec1_score          25           19             35               18   

user_name        Darshit Sabharwal  Krish Keer  Rasha Rana  Gatik Mallick  \
Tra1_Sec1_score                 18          27          27             19   
Tra1_Sec2_score                 34          29          25             33   
Tra1_Sec3_score                 31          31          22             29   
Tra1_Sec4_score                 28          31          25             29   
Tra2_Sec1_score                 26          35          32             30   

user_name        Alisha Kanda  Madhup Dasgupta  ...  Zeeshan Balasubramanian  \
Tra1_Sec1_score            26               26  ...                       89   
Tra1_Sec2_score            30               21  ...                       84   
Tra1_Sec3_score            29               30  ...                       69   
Tra1_Sec4_score            19               29  ...                       52   
Tra2_Sec1_score            25               27  ...                       43   

user_name        Kiaan Chauhan  Ira Seth  Umang Choudhary  Miraya Aurora  \
Tra1_Sec1_score             88        89               90             86   
Tra1_Sec2_score             84        86               80             86   
Tra1_Sec3_score             78        74               75             71   
Tra1_Sec4_score             55        64               53             48   
Tra2_Sec1_score             23        30               37             33   

user_name        Rhea Lad  Yakshit Brar  Rohan Chhabra  Vaibhav Borra  \
Tra1_Sec1_score        90            90             89             90   
Tra1_Sec2_score        83            86             84             83   
Tra1_Sec3_score        68            74             75             69   
Tra1_Sec4_score        52            51             57             53   
Tra2_Sec1_score        32            40             37             43   

user_name        Nayantara Yogi  
Tra1_Sec1_score              90  
Tra1_Sec2_score              88  
Tra1_Sec3_score              72  
Tra1_Sec4_score              55  
Tra2_Sec1_score              25  

[5 rows x 1005 columns]

In [122]:
# Showing best, worst, improving, deteriorating users ( *I am not able to make it dynamic*)
fig = go.Figure() 
fig.add_trace(go.Scatter(x=section_wise_score_trans.index[0:8],y=section_wise_score_trans['Dhanuk Subramanian'],
                    mode='lines+markers',
                    name='Top performer'))
fig.add_trace(go.Scatter(x=section_wise_score_trans.index[0:8],y=section_wise_score_trans['Madhav Sagar'],
                    mode='lines+markers',
                    name='worst performer'))
fig.add_trace(go.Scatter(x=section_wise_score_trans.index[0:8],y=section_wise_score_trans['Riaan Swamy'],
                    mode='lines+markers',
                    name='Improving performer'))
fig.add_trace(go.Scatter(x=section_wise_score_trans.index[0:8],y=section_wise_score_trans['Shayak Saran'],
                    mode='lines+markers',
                    name='Detoriating performer'))
fig.update_layout(title='Performers', xaxis_title='Sections', yaxis_title='Score_section_wise',height=720,width=1080)


fig.show() 

#### 2. Mentor level view - Performance of all the users for training, section, sub-section

In [19]:
# Filter the respectieve columns
# Then create summary
# Then plot it

## Training wise



In [37]:
training_wise_score.head()

user_name  Tra1_score  Tra2_score  Total_score
0         Emir Kumar         106         101          207
1        Krish Bains         105         107          212
2      Akarsh Suresh         112         126          238
3    Priyansh Bhalla         119          99          218
4  Darshit Sabharwal         111         107          218

In [254]:
# Creating Tranining wise scores Summary
data= [[training_wise_score.Tra1_score.max(),training_wise_score.Tra1_score.mean(),training_wise_score.Tra1_score.min()],[training_wise_score.Tra2_score.max(),training_wise_score.Tra2_score.mean(),training_wise_score.Tra2_score.min()]]
tra_summary = pd.DataFrame(data, columns = ['Max_score','Average_score','Min_score'])
tra_summary['Training'] = ['Training 1','Training 2']
tra_summary.insert(0, 'Trainings',tra_summary['Training'])
tra_summary.drop(columns='Training',inplace=True)
tra_summary

Trainings  Max_score  Average_score  Min_score
0  Training 1        346     201.621891         41
1  Training 2        347     201.688557         40

In [207]:
# Report: Training wise
name = ['Max_scores','Average_scores','Min_scores']
y1_ =[tra_summary.Max_scores[0],tra_summary.Average_scores[0],tra_summary.Min_scores[0]]
y2_ =[tra_summary.Max_scores[1],tra_summary.Average_scores[1],tra_summary.Min_scores[1]]
fig = go.Figure(data=[
    go.Bar(name='Training_1', x=name,y=y1_),
    go.Bar(name='Training_2', x=name,y=y2_)
  

])
# Change the bar mode
fig.update_layout(title='Report: Training wise',xaxis={'title': 'Summary'},yaxis={'title':'Total_scores'},height=520,width=720,barmode='group')
fig.show()



## Section Wise
    


In [255]:
# Creating Section wise scores Summary
data= [[section_wise_score.Tra1_Sec1_score.max(),section_wise_score.Tra1_Sec1_score.mean(),section_wise_score.Tra1_Sec1_score.min()],
[section_wise_score.Tra1_Sec2_score.max(),section_wise_score.Tra1_Sec2_score.mean(),section_wise_score.Tra1_Sec2_score.min()],
[section_wise_score.Tra1_Sec3_score.max(),section_wise_score.Tra1_Sec3_score.mean(),section_wise_score.Tra1_Sec3_score.min()],
[section_wise_score.Tra1_Sec4_score.max(),section_wise_score.Tra1_Sec4_score.mean(),section_wise_score.Tra1_Sec4_score.min()],
[section_wise_score.Tra2_Sec1_score.max(),section_wise_score.Tra2_Sec1_score.mean(),section_wise_score.Tra2_Sec1_score.min()],
[section_wise_score.Tra2_Sec2_score.max(),section_wise_score.Tra1_Sec2_score.mean(),section_wise_score.Tra2_Sec2_score.min()],
[section_wise_score.Tra2_Sec3_score.max(),section_wise_score.Tra1_Sec3_score.mean(),section_wise_score.Tra2_Sec3_score.min()],
[section_wise_score.Tra2_Sec4_score.max(),section_wise_score.Tra2_Sec4_score.mean(),section_wise_score.Tra2_Sec4_score.min()]
]
sec_summary = pd.DataFrame(data, columns = ['Max_score','Average_score','Min_score'])
sec_summary['Sections'] = ['Tra1_Section1', 'Tra1_Section2','Tra1_Section3','Tra1_Section4','Tra2_Section1','Tra2_Section2','Tra2_Section3','Tra2_Section4']
sec_summary = sec_summary[['Sections','Max_score','Average_score','Min_score']]
sec_summary

Sections  Max_score  Average_score  Min_score
0  Tra1_Section1         90      50.269652          0
1  Tra1_Section2         89      50.347264          1
2  Tra1_Section3         90      50.649751          7
3  Tra1_Section4         89      50.355224         17
4  Tra2_Section1         89      50.360199         14
5  Tra2_Section2         90      50.347264          6
6  Tra2_Section3         89      50.649751          1
7  Tra2_Section4         90      50.538308          0

In [251]:
# Report : Section Wise
name = ['Max_scores','Average_scores','Min_scores']
y1_ =[section_wise_score.Tra1_Sec1_score.max(),section_wise_score.Tra1_Sec1_score.mean(),section_wise_score.Tra1_Sec1_score.min()]
y2_ =[section_wise_score.Tra1_Sec2_score.max(),section_wise_score.Tra1_Sec2_score.mean(),section_wise_score.Tra1_Sec2_score.min()]
y3_ =[section_wise_score.Tra1_Sec3_score.max(),section_wise_score.Tra1_Sec3_score.mean(),section_wise_score.Tra1_Sec3_score.min()]
y4_ =[section_wise_score.Tra1_Sec4_score.max(),section_wise_score.Tra1_Sec4_score.mean(),section_wise_score.Tra1_Sec4_score.min()]
y5_ =[section_wise_score.Tra2_Sec1_score.max(),section_wise_score.Tra2_Sec1_score.mean(),section_wise_score.Tra2_Sec1_score.min()]
y6_ =[section_wise_score.Tra2_Sec2_score.max(),section_wise_score.Tra1_Sec2_score.mean(),section_wise_score.Tra2_Sec2_score.min()]
y7_ =[section_wise_score.Tra2_Sec3_score.max(),section_wise_score.Tra1_Sec3_score.mean(),section_wise_score.Tra2_Sec3_score.min()]
y8_ =[section_wise_score.Tra2_Sec4_score.max(),section_wise_score.Tra2_Sec4_score.mean(),section_wise_score.Tra2_Sec4_score.min()]
fig = go.Figure(data=[
    go.Bar(name='Tra1_sec1', x=name,y=y1_),
    go.Bar(name='Tra1_sec2', x=name,y=y2_),
    go.Bar(name='Tra1_sec3', x=name,y=y3_),
    go.Bar(name='Tra1_sec4', x=name,y=y4_),
    go.Bar(name='Tra2_sec1', x=name,y=y5_),
    go.Bar(name='Tra2_sec2', x=name,y=y6_),
    go.Bar(name='Tra2_sec3', x=name,y=y7_),
    go.Bar(name='Tra2_sec4', x=name,y=y8_)
  

])
# Change the bar mode
fig.update_layout(title='Report: Section wise',xaxis={'title': 'Summary'},yaxis={'title':'Total_Scores'},barmode='group',height=500,width=900)
fig.show()

## Sub-Section Wise

In [213]:
sub_section_wise.head()

Tra1_Sec1_Sub1  Tra1_Sec1_Sub2  Tra1_Sec1_Sub3  \
user_name                                                           
Emir Kumar                      4               8               6   
Krish Bains                     6               4               5   
Akarsh Suresh                   7               5               7   
Priyansh Bhalla                 5               7               9   
Darshit Sabharwal               5               2               3   

                   Tra1_Sec1_Sub4  Tra1_Sec1_Sub5  Tra2_Sec2_Sub1  \
user_name                                                           
Emir Kumar                      1               3               3   
Krish Bains                     4               9               5   
Akarsh Suresh                   2               6               7   
Priyansh Bhalla                 8               5               5   
Darshit Sabharwal               4               4               5   

                   Tra2_Sec2_Sub2  Tra2_Sec2_Sub3  Tra2_Sec2_Sub4  \
user_name                                                           
Emir Kumar                      5               4               5   
Krish Bains                     6               7               6   
Akarsh Suresh                   3               7               6   
Priyansh Bhalla                 3               6               3   
Darshit Sabharwal               3               4               7   

                   Tra2_Sec2_Sub5  total  
user_name                                 
Emir Kumar                      4     43  
Krish Bains                     4     56  
Akarsh Suresh                   6     56  
Priyansh Bhalla                 5     56  
Darshit Sabharwal               5     42

#### 3. Manager level view

In [42]:
# Top 5 performers = done
# Bottom 5 performers = done
# Performnace of all the users over time = done


### Top5 performers and Worst5 Performers

In [127]:
#Filtering data 
top5 = training_wise_score.sort_values(by='Total_score', ascending=False)
worst5 = training_wise_score.sort_values(by='Total_score', ascending=True)


In [132]:
# viz top 5 and bottom 5 for Overall

top_perf = top5.index[0:5].values
tra1 = top5.Tra1_score[0:5].values
tra2 = top5.Tra2_score[0:5].values
last_perf = worst5.index[0:5].values
tra1_ =worst5.Tra1_score[0:5].values
tra2_ =worst5.Tra2_score[0:5].values

fig = go.Figure(data=[
    go.Bar(name='Training_1', x=top_perf, y=tra1),
    go.Bar(name='Training_2', x=top_perf, y=tra2),
    go.Bar(name='Training_1', x=last_perf, y=tra1_),
    go.Bar(name='Training_2', x=last_perf, y=tra2_)
])
# Change the bar mode
fig.update_layout(title='Overall Top5 vs Worst5',xaxis={'title': 'Users'},yaxis={'title':'Scores_training wise'},barmode='group',height=520,width=720)
fig.show()


## Overall Performance of Users

Please input the number of users out of 1005 for viewing ovarall performance

In [124]:
section_wise_score

Tra1_Sec1_score  Tra1_Sec2_score  Tra1_Sec3_score  \
user_name                                                              
Emir Kumar                      22               25               31   
Krish Bains                     28               28               28   
Akarsh Suresh                   27               30               27   
Priyansh Bhalla                 34               21               38   
Darshit Sabharwal               18               34               31   
...                            ...              ...              ...   
Rhea Lad                        90               83               68   
Yakshit Brar                    90               86               74   
Rohan Chhabra                   89               84               75   
Vaibhav Borra                   90               83               69   
Nayantara Yogi                  90               88               72   

                   Tra1_Sec4_score  Tra2_Sec1_score  Tra2_Sec2_score  \
user_name                                                              
Emir Kumar                      28               25               21   
Krish Bains                     21               19               28   
Akarsh Suresh                   28               35               29   
Priyansh Bhalla                 26               18               22   
Darshit Sabharwal               28               26               24   
...                            ...              ...              ...   
Rhea Lad                        52               32               14   
Yakshit Brar                    51               40               14   
Rohan Chhabra                   57               37               13   
Vaibhav Borra                   53               43               17   
Nayantara Yogi                  55               25               17   

                   Tra2_Sec3_score  Tra2_Sec4_score  total  
user_name                                                   
Emir Kumar                      26               29    207  
Krish Bains                     24               36    212  
Akarsh Suresh                   35               27    238  
Priyansh Bhalla                 27               32    218  
Darshit Sabharwal               31               26    218  
...                            ...              ...    ...  
Rhea Lad                         9                0    348  
Yakshit Brar                     3                0    358  
Rohan Chhabra                    7                0    362  
Vaibhav Borra                    7                1    363  
Nayantara Yogi                   3                2    352  

[1005 rows x 9 columns]

In [284]:
#overall performanceof users interms of total score (Training_1 + Training_2)
print("Give the number of Users") 
n = input(int())
fig = go.Figure() 
fig.add_trace(go.Scatter(y=section_wise_score.index[0:],x=section_wise_score['total'].sort_values(ascending=True).head(int(n)),
                    mode = ('lines+markers'),
                    name='Perfomance Line'))
fig.update_layout(title='Overall performance of users',xaxis={'title': 'Total_scores'},yaxis={'title':'Users_name'},height=500,width=1000)
fig.show()

Give the number of Users
